mkosi/__init__.py

```diff
@@ -3607,13 +3634,116 @@ def insert_verity(
+    args: CommandLineArguments, root_hash: Optional[str], do_run_build_script: bool, for_cache: bool
+) -> Tuple[Optional[BinaryIO], Optional[bytes], Optional[str]]:
+
+    if do_run_build_script or args.verity != "signed":
+        return None, None, None
+    if for_cache:
+        return None, None, None
+
+    assert root_hash is not None
+    assert crypto is not None
+
+    with complete_step("Signing verity root hash…"):
+
+        key = crypto.load_privatekey(crypto.FILETYPE_PEM, args.secure_boot_key.read_text())
+        certificate = crypto.load_certificate(crypto.FILETYPE_PEM, args.secure_boot_certificate.read_text())
+
+        # Note that the library returns the SHA256 digest in an
+        # uppecase format with : as byte separators. Let's convert to
+        # classic lowercase series of unseparated hex digits
+        fingerprint = certificate.digest("sha256").decode("ascii").replace(":", "").lower()
+
+        bio_in = crypto._new_mem_buf(root_hash.encode("utf-8"))
+        pkcs7 = crypto._lib.PKCS7_sign(
+            certificate._x509,
+            key._pkey,
+            crypto._ffi.NULL,
+            bio_in,
+            PKCS7_DETACHED | PKCS7_NOCERTS | PKCS7_NOATTR | PKCS7_BINARY,
+        )
+        bio_out = crypto._new_mem_buf()
+        crypto._lib.i2d_PKCS7_bio(bio_out, pkcs7)
+        sigbytes = crypto._bio_to_string(bio_out)
+
+        b64encoded = base64.b64encode(sigbytes).decode("ascii")
+
+        # This is supposed to be extensible, but care should be taken
+        # not to include unprotected data here.
+        j = json.dumps({
+                "rootHash": root_hash,
+                "certificateFingerprint": fingerprint,
+                "signature": b64encoded
+            }).encode("utf-8")
+
+        # Pad to next multiple of 4K with NUL bytes
+        padded = j + b"\0" * (roundup(len(j), 4096) - len(j))
+
+        f: BinaryIO = cast(BinaryIO, tempfile.NamedTemporaryFile(dir=args.output.parent, prefix=".mkosi-"))
+        f.write(padded)
+        f.flush()
+
+        # Returns a file with zero-padded JSON data to insert as
+        # signature partition as first element, and the DER PKCS7
+        # signature bytes as second argument (to store as detached
+        # PKCS7 file), and finally the SHA256 fingerprint of the
+        # certificate used (which is used to deterministically
+        # generate the partition UUID for the signature partition).
+
+        return f, sigbytes, fingerprint
+
+
+def insert_verity_sig(
+    args: CommandLineArguments,
+    raw: Optional[BinaryIO],
+    loopdev: Optional[Path],
+    verity_sig: Optional[BinaryIO],
+    root_hash: Optional[str],
+    fingerprint: Optional[str],
+    for_cache: bool,
+) -> None:
+    if verity_sig is None:
+        return
+    if for_cache:
+        return
+    assert loopdev is not None
+    assert raw is not None
+    assert root_hash is not None
+    assert fingerprint is not None
+    assert args.verity_sig_partno is not None
+
```

Comment:
That seems strange. Why not remove Optional from the types instead?

Reply:
I don't follow? remove it where precisely? from the function parameter list?

That's because we want that insert_verity_sig() may be called when verity is off, and is a NOP then. i.e. the first line in this function checks for this case indirectly.

I mean, all our functions that encapsulate image building steps are put together that way: they can be called in any config, and just become nops when what they are supposed to do doesn't apply because config turned something off or something not on.

And in that case all the function args that carry the actual data are typically None, hence we must allow Optional[]. But then in the case where we actaully need to do something we have to assert that they are all set (because otherwise mypy will whine).

I mean, you could move all the conditionalizations outside of these functions, but that would mean build_image() would become super hard to follow, due to all the externalized conditionalizations. Moreover, the precise conditionalizations are often not trivial hence would suck if not encapsulated in the functions.

Or maybe I don't get at all, what you are trying to say?

Reply:
Oh, OK, I didn't look at the `is None` check above. So it's OK as is.